### ISSUE TITLE
Kubernetes cluster sometimes end up in Error state: fails to attach Binary ISO

### ISSUE THREAD

**Issue by saffronjam** [apache/cloudstack#7829]
##### ISSUE TYPE
 * Bug Report

##### COMPONENT NAME
~~~
CKS
~~~

##### CLOUDSTACK VERSION
~~~
4.18
~~~

##### CONFIGURATION
CKS 1.24.0 and 1.27.3 

(also 1.26.0, I experienced the same problems, but I did not include it in this bug report,)

I used 6 different setups using 2 service offerings with the 2 CKS-versions specified above.
Every setup had 1 controller node and the default Kubernetes isolated network offering. 

SO1 (small):
- 2 CPU cores
- 2 GB RAM
- 8 GB root disk

SO2 (big):
- 4 CPU cores
- 16 GB RAM
- 64 GB root disk

```
test-124-small

1 worker
SO 1
CKS 1.24.0
```

```
test-124-small-many

10 workers
SO 1
CKS 1.24.0
```

```
test-1273-small

1 worker
SO 1
CKS 1.27.3
```

```
test-1273-small-many

10 workers
SO 1
CKS 1.27.3
```

```
test-1273-big

1 worker
SO 2
CKS 1.27.3
```

```
test-1273-big-many

10 workers
SO 2
CKS 1.27.3
```

##### OS / ENVIRONMENT
Ubuntu 22.04 nodes running KVM

##### SUMMARY
When creating Kubernetes clusters, it fails some of the times. It appears as if it is more common when creating multiple nodes instead of just a few.

I can access the nodes in the failed clusters through ssh just fine and I am able to look at the logs, such as /var/log/daemon.log.

In the _test-1273-big-many-node1_ the entries in daemon.log indicates that the issue is related to some binary ISO not being attached.
```
Aug  8 09:30:21 systemvm cloud-init[1102]: Waiting for Binaries directory /mnt/k8sdisk/ to be available, sleeping for 15 seconds, attempt: 99
Aug  8 09:30:36 systemvm cloud-init[1102]: Waiting for Binaries directory /mnt/k8sdisk/ to be available, sleeping for 15 seconds, attempt: 100
Aug  8 09:30:51 systemvm cloud-init[1102]: Warning: Offline install timed out!

which I assume could cause the following entry:

Aug  8 09:30:52 systemvm deploy-kube-system[1420]: /opt/bin/deploy-kube-system: line 19: kubeadm: command not found
```

But that is not the case for every failed cluster, such as _test-small-many_, actually succeeds on some nodes:
```
Aug  8 09:43:10 systemvm cloud-init[1070]: Waiting for Binaries directory /mnt/k8sdisk/ to be available, sleeping for 30 seconds, attempt: 8
Aug  8 09:43:41 systemvm cloud-init[1070]: Waiting for Binaries directory /mnt/k8sdisk/ to be available, sleeping for 30 seconds, attempt: 9
Aug  8 09:44:11 systemvm cloud-init[1070]: Installing binaries from /mnt/k8sdisk/
```
But fails on the 5th node out of 10 total with the same error
```
Aug  8 10:00:16 systemvm cloud-init[1072]: Waiting for Binaries directory /mnt/k8sdisk/ to be available, sleeping for 30 seconds, attempt: 39
Aug  8 10:00:46 systemvm cloud-init[1072]: Waiting for Binaries directory /mnt/k8sdisk/ to be available, sleeping for 30 seconds, attempt: 40
Aug  8 10:01:16 systemvm cloud-init[1072]: Warning: Offline install timed out!

with the kubectl output:

➜  ~ kubectl get nodes                                      
NAME                                       STATUS   ROLES           AGE    VERSION
test-1273-small-many-control-189d4833e43   Ready    control-plane   3h7m   v1.27.3
test-1273-small-many-node-189d483a95b      Ready    <none>          3h7m   v1.27.3
test-1273-small-many-node-189d484150e      Ready    <none>          3h7m   v1.27.3
test-1273-small-many-node-189d4847014      Ready    <none>          3h7m   v1.27.3
test-1273-small-many-node-189d484b721      Ready    <none>          3h7m   v1.27.3
....
should be 6 more!
```

All logs can be found at the bottom of this issue

##### STEPS TO REPRODUCE
~~~
1. Create a cluster with one of the configurations above
2. Wait for it to be created
3. Check if it worked or not
~~~

##### EXPECTED RESULTS
~~~
Cluster ends up in Running state 
~~~

##### ACTUAL RESULTS
~~~
Cluster ends up in Error state
~~~
![dashboard](https://github.com/apache/cloudstack/assets/26722370/fda5618c-0640-4c94-9b12-0af18cd33dfa)

##### LOGS
I supplied logs from /var/log/daemon.log. I added logs from the controller and the first worker.

For the clusters with more than 1 worker, I added the logs for the first worker that failed as well.

**SUCCEEDED**
[test-124-small-controller.txt](https://github.com/apache/cloudstack/files/12291211/test-124-small-controller.txt)
[test-124-small-node1.txt](https://github.com/apache/cloudstack/files/12291215/test-124-small-node1.txt)

**FAILED**
[test-124-small-many-controller.txt](https://github.com/apache/cloudstack/files/12291212/test-124-small-many-controller.txt)
[test-124-small-many-node1.txt](https://github.com/apache/cloudstack/files/12291214/test-124-small-many-node1.txt)
[test-124-small-many-node2.txt](https://github.com/apache/cloudstack/files/12291652/test-124-small-many-node2.txt)

**FAILED**
[test-1273-small-controller.txt](https://github.com/apache/cloudstack/files/12291221/test-1273-small-controller.txt)
[test-1273-small-node1.txt](https://github.com/apache/cloudstack/files/12291224/test-1273-small-node1.txt)

**FAILED**
[test-1273-small-many-controller.txt](https://github.com/apache/cloudstack/files/12291222/test-1273-small-many-controller.txt)
[test-1273-small-many-node1.txt](https://github.com/apache/cloudstack/files/12291223/test-1273-small-many-node1.txt)
[test-1273-small-many-node5.txt](https://github.com/apache/cloudstack/files/12291554/test-1273-small-many-node5.txt)

**SUCCEEDED**
[test-1273-big-controller.txt](https://github.com/apache/cloudstack/files/12291216/test-1273-big-controller.txt)
[test-1273-big-node1.txt](https://github.com/apache/cloudstack/files/12291219/test-1273-big-node1.txt)

**FAILED**
[test-1273-big-many-controller.txt](https://github.com/apache/cloudstack/files/12291217/test-1273-big-many-controller.txt)
[test-1273-big-many-node1.txt](https://github.com/apache/cloudstack/files/12291218/test-1273-big-many-node1.txt)
(failed on the first node)

**Comment by saffronjam** [comment]
Sorry for leaving that information out.

We have set that to **50**

**Comment by weizhouapache** [comment]
@saffronjam 
do you use local storage or shared storage ?

**Comment by saffronjam** [comment]
We use this:

**Storage type**
Shared (ROOT)

**Comment by saffronjam** [comment]
Secondary storage is also NFS, yes.

We are in a bit of a transition state, where we have 3 secondary storage, but only **1** of them is read/write, the other 2 are read only

To add to this, the 1.27.3 Kuberenetes ISO (that was used in my tests) was added after we put 2 secondary storage is read only mode, so it should only be in one of the secondary storages.

**Comment by weizhouapache** [comment]
> Secondary storage is also NFS, yes.
> 
> We are in a bit of a transition state, where we have 3 secondary storage, but only **1** of them is read/write, the other 2 are read only
> 
> To add to this, the 1.27.3 Kuberenetes ISO (that was used in my tests) was added after we put 2 secondary storage is read only mode, so it should only be in one of the secondary storages.

@saffronjam 
can you test the changes in #7848 ?

**Comment by weizhouapache** [comment]
> We don't have a test-setup running, is there any guide or instructions on how to we can do this?
> 
> We are running a management server with a database on a single node, then agents on a few other nodes. All installed with apt.
> 
> Do we shutdown the current management server, build your changes into jar-files and then try to start that using the same settings and database etc.?

@saffronjam 
if you use 4.18.0.0, here are the steps
```
// on any other server
git clone https://github.com/apache/cloudstack
cd cloudstack
git checkout 4.18.0.0
wget https://raw.githubusercontent.com/apache/cloudstack/737ac561f8b6dd1cefd17c2fdf82b5846aea20bf/server/src/main/java/com/cloud/template/TemplateManagerImpl.java -O server/src/main/java/com/cloud/template/TemplateManagerImpl.java
mvn -P developer,systemvm clean install
// copy client/target/cloud-client-ui-4.18.0.0.jar to /usr/share/cloudstack-management/lib/cloudstack-4.18.0.0.jar on management server (backup at first)
// restart cloudstack-management service on management server
```

**Comment by saffronjam** [comment]
Perfect, we are running some tests right now. I'll get back to you tonight or tomorrow morning.

**Comment by weizhouapache** [comment]
> Perfect, we are running some tests right now. I'll get back to you tonight or tomorrow morning.

@saffronjam good. please keep updated, thanks

**Comment by saffronjam** [comment]
The issue appears to remain. 

![image](https://github.com/apache/cloudstack/assets/26722370/667675a5-2bdc-417c-b6ce-6fa9fe523657)

We tried CKS 1.24 and 1.27.3 (the one without version in the name) with 7 and 10 nodes respectively. I tried with and without HA (2 control nodes, made no difference). The last test was with HA, so that's what I'm reporting on.

1.24:
```
➜  ~ kubectl get nodes
NAME                                   STATUS   ROLES           AGE   VERSION
test-weizhou-124-control-189e015f8ad   Ready    control-plane   17h   v1.24.0
```
[test-weizhou-124-control2.txt](https://github.com/apache/cloudstack/files/12320159/test-weizhou-124-control2.txt)

1.27.3:
```
➜  ~ kubectl get nodes
NAME                               STATUS   ROLES           AGE   VERSION
test-weizhou-control-189e0152f9d   Ready    control-plane   17h   v1.27.3
test-weizhou-control-189e0156d4a   Ready    control-plane   17h   v1.27.3
test-weizhou-node-189e015b3a9      Ready    <none>          17h   v1.27.3
test-weizhou-node-189e015ead7      Ready    <none>          17h   v1.27.3
test-weizhou-node-189e0162b2e      Ready    <none>          17h   v1.27.3
```
[test-weizhou-node4.txt](https://github.com/apache/cloudstack/files/12319995/test-weizhou-node4.txt)

The result is pretty much same as earlier where it get stuck on some node, as you can read in the logs.
```
/var/log/daemon.log

Aug 10 16:00:30 systemvm cloud-init[1109]: Waiting for Binaries directory /mnt/k8sdisk/ to be available, sleeping for 15 seconds, attempt: 100
Aug 10 16:00:45 systemvm cloud-init[1109]: Warning: Offline install timed out!
```

Lmk if you need more information about our setup

**Comment by weizhouapache** [comment]
> The issue appears to remain.
> 
> ![image](https://user-images.githubusercontent.com/26722370/259963389-667675a5-2bdc-417c-b6ce-6fa9fe523657.png)
> 
> We tried CKS 1.24 and 1.27.3 (the one without version in the name) with 7 and 10 nodes respectively. I tried with and without HA (2 control nodes, made no difference). The last test was with HA, so that's what I'm reporting on.
> 
> 1.24:
> 
> ```
> ➜  ~ kubectl get nodes
> NAME                                   STATUS   ROLES           AGE   VERSION
> test-weizhou-124-control-189e015f8ad   Ready    control-plane   17h   v1.24.0
> ```
> 
> [test-weizhou-124-control2.txt](https://github.com/apache/cloudstack/files/12320159/test-weizhou-124-control2.txt)
> 
> 1.27.3:
> 
> ```
> ➜  ~ kubectl get nodes
> NAME                               STATUS   ROLES           AGE   VERSION
> test-weizhou-control-189e0152f9d   Ready    control-plane   17h   v1.27.3
> test-weizhou-control-189e0156d4a   Ready    control-plane   17h   v1.27.3
> test-weizhou-node-189e015b3a9      Ready    <none>          17h   v1.27.3
> test-weizhou-node-189e015ead7      Ready    <none>          17h   v1.27.3
> test-weizhou-node-189e0162b2e      Ready    <none>          17h   v1.27.3
> ```
> 
> [test-weizhou-node4.txt](https://github.com/apache/cloudstack/files/12319995/test-weizhou-node4.txt)
> 
> The result is pretty much same as earlier where it get stuck on some node, as you can read in the logs.
> 
> ```
> /var/log/daemon.log
> 
> Aug 10 16:00:30 systemvm cloud-init[1109]: Waiting for Binaries directory /mnt/k8sdisk/ to be available, sleeping for 15 seconds, attempt: 100
> Aug 10 16:00:45 systemvm cloud-init[1109]: Warning: Offline install timed out!
> ```
> 
> Lmk if you need more information about our setup

@saffronjam 
ok. can you go to the kvm hosts and check if the k8s ISO is attached to the control nodes and worker nodes ?
if ISO is attached, please also check if the file path  exists or not.

**Comment by saffronjam** [comment]
@weizhouapache I checked /mnt on the control and worker nodes, but nothing is mounted. /mnt is completely empty on every control/worker node I checked.  

How do I check attached ISO from the KVM hosts? Is it some virsh command?

**Comment by weizhouapache** [comment]
> @weizhouapache I checked /mnt on the control and worker nodes, but nothing is mounted. /mnt is completely empty on every control/worker node I checked.
> 
> How do I check attached ISO from the KVM hosts? Is it some virsh command?

@saffronjam can you check management-server.log if there are following logs for the not-working nodes ?
```
2023-08-10 07:19:51,894 DEBUG [c.c.a.t.Request] (API-Job-Executor-1:ctx-411c8566 job-35 ctx-605823c2) (logid:920e9c8d) Seq 1-7322571519127715877: Sending  { Cmd , MgmtId: 167780772, via: 1(ref-trl-5430-k-Mr8-wei-zhou-kvm1), Ver: v1, Flags: 100011, [{"org.apache.cloudstack.storage.command.AttachCommand":{"disk":{"data":{"org.apache.cloudstack.storage.to.TemplateObjectTO":{"path":"template/tmpl/1/202/202-1-df500cdb-7e18-328d-bee0-65ef80fff244.iso","origUrl":"http://10.0.3.122/cks/setup-1.26.0.iso","uuid":"68f04b1f-ab7e-444e-a1f5-57d054e6b4ee","id":"202","format":"ISO","accountId":"1","checksum":"{SHA-512}ad026d55e0fda7dcb65764660258084f9985d8800b684228a367cf7f90a44d05fb3ff83d9d4e6bc0da2b7049c8bdbfa8c432b342faf96386f92666bcbec8a216","hvm":"true","displayText":"v1.26.0-Kubernetes-Binaries-ISO","imageDataStore":{"com.cloud.agent.api.to.NfsTO":{"_url":"NFS://10.0.32.4/acs/secondary/ref-trl-5430-k-Mr8-wei-zhou/ref-trl-5430-k-Mr8-wei-zhou-sec1","_role":"Image"}},"name":"202-1-df500cdb-7e18-328d-bee0-65ef80fff244","size":"(660.03 MB) 692094976","hypervisorType":"None","bootable":"false","uniqueName":"202-1-df500cdb-7e18-328d-bee0-65ef80fff244","directDownload":"false","deployAsIs":"false"}},"type":"ISO"},"vmName":"i-2-3-VM","inSeq":"false","forced":"true","controllerInfo":{"cpuOvercommitRatio":"2.0","memoryOvercommitRatio":"1.0","rootdisksize":"10"},"wait":"0","bypassHostMaintenance":"false"}}] }
2023-08-10 07:19:52,083 DEBUG [c.c.a.t.Request] (AgentManager-Handler-17:null) (logid:) Seq 1-7322571519127715877: Processing:  { Ans: , MgmtId: 167780772, via: 1, Ver: v1, Flags: 10, [{"com.cloud.agent.api.Answer":{"result":"true","wait":"0","bypassHostMaintenance":"false"}}] }
2023-08-10 07:19:52,084 DEBUG [c.c.a.t.Request] (API-Job-Executor-1:ctx-411c8566 job-35 ctx-605823c2) (logid:920e9c8d) Seq 1-7322571519127715877: Received:  { Ans: , MgmtId: 167780772, via: 1(ref-trl-5430-k-Mr8-wei-zhou-kvm1), Ver: v1, Flags: 10, { Answer } }
2023-08-10 07:19:52,100 INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-1:ctx-411c8566 job-35 ctx-605823c2) (logid:920e9c8d) Attached binaries ISO for VM : advanced-sg-control-189de4fe0e7 in cluster: advanced-sg
```

**Comment by saffronjam** [comment]
I created a new cluster to test this with 1 + 10 nodes. In this cluster, 

1. I can find the "org.apache.cloudstack.storage.command.AttachCommand" entries for every node. **Even the failed ones**

2. The logs and the CloudStack GUI reports the jobs being finished (both attach and detach jobs):
```
Aug 14 12:27:37 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Attached binaries ISO for VM : test-cluster-control-189f3756fa2 in cluster: test-cluster
Aug 14 12:27:37 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Attached binaries ISO for VM : test-cluster-node-189f375d12e in cluster: test-cluster
... same for the nodes in between
Aug 14 12:27:43 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Attached binaries ISO for VM : test-cluster-node-189f3787958 in cluster: test-cluster
Aug 14 12:27:43 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Attached binaries ISO for VM : test-cluster-node-189f378ccf0 in cluster: test-cluster
```
```
Aug 14 12:56:11 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Detached Kubernetes binaries from VM : test-cluster-control-189f3756fa2 in the Kubernetes cluster : test-cluster
Aug 14 12:56:12 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Detached Kubernetes binaries from VM : test-cluster-node-189f375d12e in the Kubernetes cluster : test-cluster
... same for the nodes in between
Aug 14 12:56:54 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Detached Kubernetes binaries from VM : test-cluster-node-189f3787958 in the Kubernetes cluster : test-cluster
Aug 14 12:56:56 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Detached Kubernetes binaries from VM : test-cluster-node-189f378ccf0 in the Kubernetes cluster : test-cluster
```

I've tried to find logs after the Attach and Detach jobs completes, but I only find one log entry after the last detach saying that the job failed:

```
2023-08-14 12:56:56,831 INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Detached Kubernetes binaries from VM : test-cluster-node-189f378ccf0 in the Kubernetes cluster : test-cluster
2023-08-14 12:56:56,832 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-14:ctx-63b2bc93 job-26834) (logid:bc9fab70) Complete async job-26834, jobStatus: FAILED, resultCode: 530, result: org.apache.cloudstack.api.response.ExceptionResponse/null/{"uuidList":[],"errorcode":"530","errortext":"Failed to setup Kubernetes cluster : test-cluster in usable state as unable to provision API endpoint for the cluster"}
```

**Comment by zap51** [comment]
> I created a new cluster to test this with 1 + 10 nodes. In this cluster, 
> 
> 1. I can find the "org.apache.cloudstack.storage.command.AttachCommand" entries for every node. **Even the failed ones**
> 
> 2. The logs and the CloudStack GUI reports the jobs being finished (both attach and detach jobs):
> ```
> Aug 14 12:27:37 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Attached binaries ISO for VM : test-cluster-control-189f3756fa2 in cluster: test-cluster
> Aug 14 12:27:37 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Attached binaries ISO for VM : test-cluster-node-189f375d12e in cluster: test-cluster
> ... same for the nodes in between
> Aug 14 12:27:43 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Attached binaries ISO for VM : test-cluster-node-189f3787958 in cluster: test-cluster
> Aug 14 12:27:43 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Attached binaries ISO for VM : test-cluster-node-189f378ccf0 in cluster: test-cluster
> ```
> ```
> Aug 14 12:56:11 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Detached Kubernetes binaries from VM : test-cluster-control-189f3756fa2 in the Kubernetes cluster : test-cluster
> Aug 14 12:56:12 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Detached Kubernetes binaries from VM : test-cluster-node-189f375d12e in the Kubernetes cluster : test-cluster
> ... same for the nodes in between
> Aug 14 12:56:54 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Detached Kubernetes binaries from VM : test-cluster-node-189f3787958 in the Kubernetes cluster : test-cluster
> Aug 14 12:56:56 se-flem-001 java[595738]: INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Detached Kubernetes binaries from VM : test-cluster-node-189f378ccf0 in the Kubernetes cluster : test-cluster
> ```
> 
> I've tried to find relevant logs after the Attach and Detach jobs completes, but I only find one log entry after the last detach saying that the job failed:
> 
> ```
> 2023-08-14 12:56:56,831 INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-14:ctx-63b2bc93 job-26834 ctx-15a4fa84) (logid:bc9fab70) Detached Kubernetes binaries from VM : test-cluster-node-189f378ccf0 in the Kubernetes cluster : test-cluster
> 2023-08-14 12:56:56,832 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-14:ctx-63b2bc93 job-26834) (logid:bc9fab70) Complete async job-26834, jobStatus: FAILED, resultCode: 530, result: org.apache.cloudstack.api.response.ExceptionResponse/null/{"uuidList":[],"errorcode":"530","errortext":"Failed to setup Kubernetes cluster : test-cluster in usable state as unable to provision API endpoint for the cluster"}
> ```
> 
> So it seems as we get a timeout  (_Waiting for Binaries directory /mnt/k8sdisk/ to be available, sleeping for 15 seconds, attempt: 100_) even if the logs says the attach/detach jobs runs

I'll try to reproduce this issue and come back. @weizhouapache has provided the patch to only consider images / ISOs if the secondary storage has the image in READY state but it still doesn't work.

@saffronjam would you be able to check if the binary ISOs present on the secondary storage match on all the NFS shares (RW and RO) , perhaps a checksum would be good. Also, do we happen to see any network issues by any chance between the hypervisors and the NFS server? I haven't took a total look on the logs, I'll come back.

Thanks,
Jayanth

**Comment by weizhouapache** [comment]
> I'll try to reproduce this issue and come back. @weizhouapache has provided the patch to only consider images / ISOs if the secondary storage has the image in READY state but it still doesn't work.
> 
> @saffronjam would you be able to check if the binary ISOs present on the secondary storage match on all the NFS shares (RW and RO), perhaps a checksum would be good. Also see if you can send the output of `# journalctl -u setup-kube-system.service --no-pager` if feasible.
> 
> Also, do we happen to see any network issues by any chance between the hypervisors and the NFS server? I haven't took a total look on the logs, I'll come back. As far as I see, this has something to do with the binary images but not sure.
> 
> Thanks, Jayanth

@zap51 
good point, it might be possible that the ISO is Ready (CloudStack state) on the secondary storages but actually not

**Comment by zap51** [comment]
> > I'll try to reproduce this issue and come back. @weizhouapache has provided the patch to only consider images / ISOs if the secondary storage has the image in READY state but it still doesn't work.
> > @saffronjam would you be able to check if the binary ISOs present on the secondary storage match on all the NFS shares (RW and RO), perhaps a checksum would be good. Also see if you can send the output of `# journalctl -u setup-kube-system.service --no-pager` if feasible.
> > Also, do we happen to see any network issues by any chance between the hypervisors and the NFS server? I haven't took a total look on the logs, I'll come back. As far as I see, this has something to do with the binary images but not sure.
> > Thanks, Jayanth
> 
> @zap51 good point, it might be possible that the ISO is Ready (CloudStack state) on the secondary storages but actually not

@weizhouapache The scenarios are as follows

1. The secondary storage is totally unavailable where the NFS client can't even mount it. In this case it appears, the exception is thrown

```
Failed to attach binaries ISO for VM : cks-many-node-189f864d00a in the Kubernetes cluster name: cks-many
``` 

So, if the hypervisor can't mount the share which should be something like the below, the above exception is thrown.

```
10.231.15.201:/secondary/template/tmpl/1/213 nfs4      17T  5.5G   17T   1% /mnt/59c5727a-6a4a-32c0-9e27-fc300161f497
```

2. In some cases (where I've seen) NFS server becomes buggy and doesn't accept reads or writes but mount operations go through might end up in this issue.

3. If there are multiple secondary storage servers, below could be some possible scenarios
  - The secondary storage (RW) might be experiencing issues like in point (2).
  - The secondary storage (RO) doesn't have the ISO. The directory under `template/tmpl/1/<id>` where`<id>` should have not got created but this should end up in point (1).
  - The secondary storage (RO) has the directory structure created `template/tmpl/1/<id>` but there is no ISO or there is partial ISO named something `_tmp`. In this case, the mount should have gone through but the ISO can not be attached. Does it throw the same exception as point (1)?

5. @weizhouapache How does ACS find which secondary storage to mount? Does it mount random secondary storage servers or has some order?

@saffronjam Maybe we can get started by finding which secondary storage has the ISOs present. This should be found under `template/tmpl/1/<id>`. The `template.properties` under each `<id>` file should have the description of the template. 
During the reproducing of the issue, please check which of your NFS servers are being mounted on the hypervisors using something like `$ df -Th | grep nfs` and also simultaneously try checking your worker / master nodes failing to mount the ISOs,

```
ssh -p <ssh_port> -o stricthostkeychecking=off cloud@<Ingress_IP> 'sudo grep "/mnt/k8*" /var/log/daemon.log'
```

Also please share the management server logs.

Thanks,
Jayanth Reddy

**Comment by weizhouapache** [comment]
> 
> 4. @weizhouapache How does ACS find which secondary storage to mount? Does it mount random secondary storage servers or has some order?
> 

@zap51 
ACS randomly select a secondary storage and mount the ISO on it to the k8s nodes.
from what @saffronjam shared, the ISO has been attached to k8s nodes successfully, but it works on some nodes and does not work on others.
maybe better to double check if the working/failed nodes use the same ISO.

**Comment by weizhouapache** [comment]
> > > 4. @weizhouapache How does ACS find which secondary storage to mount? Does it mount random secondary storage servers or has some order?
> > 
> > 
> > @zap51
> > ACS randomly select a secondary storage and mount the ISO on it to the k8s nodes.
> > from what @saffronjam shared, the ISO has been attached to k8s nodes successfully, but it works on some nodes and does not work on others.
> > maybe better to double check if the working/failed nodes use the same ISO.
> 
> @weizhouapache Right, thanks! The ISO should be `/dev/sr0` on the master and worker nodes then. @saffronjam would you also please confirm if the device of type `iso*` is available on the worker and master nodes that are failing to find binaries? I doubt if the device itself is available on the nodes.
> 
> Thanks, Jayanth

CKS gets the device name of the attached ISO by command `blkid -o device -t TYPE=iso9660` (run `sudo -i` at first)
It is possible that the ISO is mounted but the above command return empty result.
would be good to check by `blkid -o device -t LABEL=CDROM` (see #7650 )

**Comment by saffronjam** [comment]
Hi!
Thanks for the help, I've tried the things you suggested.

1. I have checked each secondary storage for the ISOs I have used in my tests.

Only the RW-storage (1 of 3) has the 1.27.3 Kubernetes ISO (the ID is 275 just for reference). 
Only the RW and one of the RO storage has 1.24.

I also checked the MD5 sums and everything is in order.

2. As I understand it, the ISO get mounted temporarily on the hypervisor as long as some cluster node needs it?
I found that the correct secondary storage is mounted with the correct ISO, for example, the RW secondary storage with ISO with ID 275.

```
nfs.cloud.cbh.kth.se:/mnt/cloud/cloudstack/sec/template/tmpl/1/275 nfs4    39T   21G   39T   1% /mnt/b82f3155-51f4-3fa2-b7ff-2f3ed76e7585
```

**However**, for some hypervisors I was unable to check this since the command `df -Th | grep nfs` got stuck (gave me no output and forced me to CTRL + C), same with `ls /mnt`. This coincidentally correlates to which cluster node that either fails because of too many attempts to wait for /mnt/k8sdisk/ or just takes longer time to mount it (the node was se-flem-016 just for reference). 

I tried to reboot the 016-hypervisor and I am now able to run `df -Th | grep nfs`. I will try to reproduce this problem again and see if it correlates. 

@weizhouapache This leaves me with a question though. If the binaries fails to attach for any reason, how do we "reset" that cluster node so it tries to attach the ISO again?

**Comment by saffronjam** [comment]
Quick update:
I am able to reproduce this issue. Whenever a cluster node is started on a hypervisor that is can't run `df -Th | grep nfs` as described above, then, quite unsurprisingly, the K8s ISO will not mount. I will try to find the issue with mounting.

**Comment by weizhouapache** [comment]
@saffronjam 
maybe check your NFS configurations to see if the ISO can be accessed by the host ?

**Comment by zap51** [comment]
@saffronjam 
> **However**, for some hypervisors I was unable to check this since the command `df -Th | grep nfs` got stuck (gave me no output and forced me to CTRL + C), same with `ls /mnt`. This coincidentally correlates to which cluster node that either fails because of too many attempts to wait for /mnt/k8sdisk/ or just takes longer time to mount it (the node was se-flem-016 just for reference).

Just like I expected. Sometimes it is hard to deal with NFS.

> I tried to reboot the 016-hypervisor and I am now able to run `df -Th | grep nfs.`

You should have done `# umount -l /mnt` instead of a reboot and checked the problem instead.


> @weizhouapache This leaves me with a question though. If the binaries fails to attach for any reason, how do we "reset" that cluster node so it tries to attach the ISO again?

Please check if you can try running `# systemctl start setup-kube-system && systemctl start deploy-kube-system`. The former should attempt to re-attach the ISO and the latter will proceed with the deployment.

> Quick update:
> I was able to reproduce the issue, and the failing node was run on a hypervisor that can't run the command df -Th | grep nfs as described above. I will try to find some logs here.

Please inspect `# dmesg -T` or `rsyslog`. You'll find hints there.

Thanks,
Jayanth

**Comment by saffronjam** [comment]
Hi,

I believe we have found the issue, and I would just like to report what happened.

In the tests I used 1.24.0 and 1.27.3 Kubernetes versions.

These existed in two secondary storages such as:

- Storage 1: **1.24.0**
- Storage 2: **1.24.0** & **1.27.3**

The issue was the Storage 1 was behaving incorrectly, either cause of permissions or networking issues, which caused a node to the stuck in the mounting process if it decided to use that storage (which according to @weizhouapache above was a random process)

So, when I tried to create clusters with 1.24.0 in my tests, those hypervisors that decided to bind the 1.24.0 from storage 1 would get stuck and therefore, until restart or unmounted, any other cluster that needed the binaries to mount on that hypervisor could fail.  Hence the reason it could fail on other version that did not exist on the faulty secondary storage.

The solution was to remove the storage and make sure to unmount it on every hypervisor.

Now, I do believe this is users' fault and that NFS should be assumed to be correctly configured, but perhaps we should look over some sort solution where I could have been informed, such as a log entry in the management server "Failed to mount ISO on hypervisor. Server <address> might be misbehaving". 

Anyhow, we can close this issue, and I'd like to thank you guys for all the help. While it was not an issue with CloudStack, it gave a great insight to the internal machinery and we'll be able to troubleshoot much easier in the future.
